README.md
# Nats::Pain

Welcome to your new gem! In this directory, you'll find the files you need to be able to package up your Ruby library into a gem. Put your Ruby code in the file `lib/nats/pain`. To experiment with that code, run `bin/console` for an interactive prompt.

TODO: Delete this and the text above, and describe your gem

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'nats-pain'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install nats-pain

## Usage

This gem simplifies usage of nats and nats-streaming in pair.

In Iconjob company we have created a very exact delivering nats message system.

Firstly it tries to send message with Nats-streaming, then with regular nats.
All messages have their own transactions. 

So consuming services would not make any duplications. **It's strongly recommended to be implemented!**

### Environment Variables
* PAIN_NATS_ENABLED - is nats enabled or not. Default: `true`
* PAIN_NATS_POOL_SIZE - size of the connection pool. Default: `5`
* PAIN_NATS_POOL_TIMEOUT - timeout in seconds for pool. Default: `5`

* PAIN_STAN_ENABLED - is stan enabled or not. Default: `true`
* PAIN_STAN_POOL_SIZE - size of the connection pool. Default: `5`
* PAIN_STAN_POOL_TIMEOUT - timeout in seconds for pool. Default: `5`

* NATS_SERVICE_NAME - name of service that is using this gem. Default: `painful_service`
* NATS_SERVERS_URLS - array of nats servers to connect with. Default: `nats://localhost:4222`
* STAN_SERVERS_URLS - array of stan servers to connect with. Default: `nats://localhost:4223`
* STAN_CLUSTER_NAME - name of stan cluster to connect with. Default: `worki_cluster`

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/[USERNAME]/nats-pain. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the Nats::Pain project’s codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/[USERNAME]/nats-pain/blob/master/CODE_OF_CONDUCT.md).


CHANGELOG.md
# Version 0.3.0

Updated hash structure that could be used to load stan/nats services. 

# Version 0.2.2

Dropped all dry-configurable mentions

# Version 0.2.1

Dropped dry-configurable because of the conflicts for old RoR projects

# Version 0.2.0

Refactored to a connection_pool for each nats/stan client.

# Version 0.1.2

Refactored a bit connection to use env variables with dry-configurable.

# Version 0.1.1

Added readers for NATS and STAN connections.

# Version 0.1.0

Initial release that's connected to `nats_listener` and `nats_streaming_listener` gems.

spec/spec_helper.rb
require "bundler/setup"
require "nats/pain"

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/nats/pain_spec.rb
RSpec.describe Nats::Pain do
  it 'has a version number' do
    expect(Nats::Pain::VERSION).not_to be nil
  end
end


spec/nats/pain/connection_spec.rb
RSpec.describe Nats::Pain::Connection do
  describe '.new' do
    let(:opts) { {} }
    subject { described_class.new(opts) }

    it 'should create connection' do
      expect { subject }.not_to raise_exception
    end
  end

  describe '.current' do
    subject { described_class.current }
    let(:topic) { 'topic' }
    let(:message) { '{"a": "123"}' }

    it 'should return connection pool' do
      expect(subject).to be_an_instance_of(described_class)
    end

    it 'should publish nats event' do
      expect_any_instance_of(NatsListener::Client).to receive(:publish).with(topic, message).and_return(true)
      subject.nats_pool&.with { |conn| conn.publish(topic, message) }
    end

    describe 'with disabled nats' do
      let(:opts) { { nats_enabled: false } }
      it 'should not fail' do
        expect { subject.nats_pool&.with { |conn| conn.publish(topic, message) } }.not_to raise_exception
      end
    end
  end
end


lib/nats/pain/connection.rb
require 'securerandom'
require 'nats_listener'
require 'nats_streaming_listener'
require 'ougai'
require 'connection_pool'

module Nats
  module Pain
    class Connection
      DEFAULT_NATS_SERVERS = 'nats://localhost:4222'.freeze
      DEFAULT_STAN_SERVERS = 'nats://localhost:4223'.freeze
      DEFAULT_SERVICE_NAME = 'painful_service'.freeze
      DEFAULT_STAN_CLUSTER_NAME = 'worki_cluster'.freeze

      attr_reader :nats_pool, :stan_pool

      def self.current
        @current ||= Nats::Pain::Connection.new
      end

      class << self
        attr_writer :current
      end

      # Structure of opts
      # @param [Hash] opts options to create connection
      # @option opts [Logger] :logger logger for connection
      # @option [Hash] :nats the options to be passed to nats listener.
      #   @option nats [Boolean] :enabled - flag to start nats or not
      #   @option nats [Integer] :pool_size - pool size
      #   @option nats [Integer] :pool_timeout - pool timeout
      #   @option nats [Boolean] :catch_errors - flag for catching errors
      #   @option nats [class] :catch_provider this class will be called with catch_provider.error(e)
      #   @option nats [String] :service_name - service name
      #   @option nats [Hash] :nats - nats-pure config for establishing connection
      # @option [Hash] :stan Hash for stan listener
      #   @option stan [Boolean] :enabled - flag to start stan or not
      #   @option stan [Integer] :pool_size - pool size
      #   @option stan [Integer] :pool_timeout - pool timeout
      #   @option stan [Boolean] :catch_errors - flag for catching errors
      #   @option stan [class] :catch_provider this class will be called with catch_provider.error(e)
      #   @option stan [String] :cluster_name - name of nats-streaming cluster that we connect to
      #   @option stan [String] :nats - nats connection info(example: ```{servers: 'nats://127.0.0.1:4223'}```)
      #   @option stan [String] :service_name - name of current service
      #   @option stan [String] :client_id - current service client id(optional)

      def initialize(opts = {})
        @logger = opts.fetch(:logger) { Ougai::Logger.new(STDOUT) }

        load_nats(opts.fetch(:nats) { {} })
        load_stan(opts.fetch(:stan) { {} })
      end

      def load_nats(opts = {})
        return unless opts.fetch(:enabled) { true }

        @nats_pool = ConnectionPool.new(size: opts.fetch(:pool_size) { 5 }, timeout: opts.fetch(:pool_timeout) { 5 }) do
          connection = NatsListener::Client.new(
            logger: @logger,
            skip: false,
            catch_errors: opts.fetch(:catch_errors) { false },
            catch_provider: opts.fetch(:catch_provider) { nil }
          )
          nats_config = opts.fetch(:nats) { { servers: opts.fetch(:nats_servers) { DEFAULT_NATS_SERVERS }.split(',') } }
          connection.establish_connection(
            service_name: opts.fetch(:service_name) { DEFAULT_SERVICE_NAME },
            nats: nats_config, # Options passed to nats connector
          )
          connection
        end
      end

      def load_stan(opts = {})
        return unless opts.fetch(:enabled) { true }

        @stan_pool = ConnectionPool.new(size: opts.fetch(:pool_size) { 5 }, timeout: opts.fetch(:pool_timeout) { 5 }) do
          connection = NatsStreamingListener::StreamingClient.new(
            logger: @logger,
            skip: false,
            catch_errors: opts.fetch(:catch_errors) { false },
            catch_provider: opts.fetch(:catch_provider) { nil }
          )
          nats_config = { servers: opts.fetch(:stan_servers) { DEFAULT_STAN_SERVERS }.split(',') }
          connection.establish_connection(
            service_name: opts.fetch(:nats_service_name) { DEFAULT_SERVICE_NAME },
            nats: opts.fetch(:nats) { nats_config }, # Options passed to nats connector
            cluster_name: opts.fetch(:stan_cluster_name) { DEFAULT_STAN_CLUSTER_NAME }, # Cluster of nats-streaming that you're connecting to
            client_id: opts.fetch(:client_id) { SecureRandom.hex(13).to_s } # Id of a client(nats-streaming works with unique client_id)
          )
          connection
        end
      end
    end
  end
end
